tests/test_defects.py

```diff
@@ -0,0 +1,270 @@
+        # x, y, size tuples
+        # always put edge defects at the start and change the value of nEdge
+        self.brightDefects = [(0, 15, 3), (100, 123, 1), (77, 90, 3)]
+        self.darkDefects = [(25, 0, 1), (33, 62, 2), (52, 21, 2)]
+        nEdge = 1  # NB update if more edge defects are included
+        self.noEdges = slice(nEdge, None)
+        self.onlyEdges = slice(0, nEdge)
+
+        self.darkBBoxes = [Box2I(Point2I(x, y), Extent2I(s, s)) for (x, y, s) in self.darkDefects]
+        self.brightBBoxes = [Box2I(Point2I(x, y), Extent2I(s, s)) for (x, y, s) in self.brightDefects]
+
+        flatWidth = np.sqrt(self.flatMean) + self.readNoiseAdu
+        darkWidth = self.readNoiseAdu
+        self.rng = np.random.RandomState(0)
+        flatData = self.rng.normal(self.flatMean, flatWidth, (shapeX, shapeY))
+        darkData = self.rng.normal(self.darkMean, darkWidth, (shapeX, shapeY))
+
+        # NB darks and flats have same defects applied deliberately to both
+        for defect in self.brightDefects:
+            y, x, s = defect
+            flatData[x:x+s, y:y+s] += self.nSigmaBright * flatWidth  # are these actually the numbers we want?
+            darkData[x:x+s, y:y+s] += self.nSigmaBright * darkWidth  # are these actually the numbers we want?
+
+        for defect in self.darkDefects:
+            y, x, s = defect
+            flatData[x:x+s, y:y+s] -= self.nSigmaDark * flatWidth  # are these actually the numbers we want?
+            darkData[x:x+s, y:y+s] -= self.nSigmaDark * darkWidth  # are these actually the numbers we want?
+
+        self.darkExp = self.flatExp.clone()
+        self.spareImage = self.flatExp.clone()  # for testing edge bits and misc
+
+        self.flatExp.image.array[:] = flatData
+        self.darkExp.image.array[:] = darkData
+
+        self.defaultTask = cpPipe.defects.FindDefectsTask(config=self.defaultConfig)
+
+    def tearDown(self):
+        del self
+
+    def test_defectFindingAllSensor(self):
+        config = copy.copy(self.defaultConfig)
+        config.nPixBorderUpDown = 0
+        config.nPixBorderLeftRight = 0
+        task = cpPipe.defects.FindDefectsTask(config=config)
+
+        defects = task.findHotAndColdPixels(self.flatExp, 'flat')
+
+        allBBoxes = self.darkBBoxes + self.brightBBoxes
+
+        for defect in defects:
+            self.assertIn(defect.getBBox(), allBBoxes)
+
+    def test_defectFindingEdgeIgnore(self):
+        task = cpPipe.defects.FindDefectsTask(config=self.defaultConfig)
+        defects = task.findHotAndColdPixels(self.flatExp, 'flat')
+
+        shouldBeFound = self.darkBBoxes[self.noEdges] + self.brightBBoxes[self.noEdges]
+        for defect in defects:
+            self.assertIn(defect.getBBox(), shouldBeFound)
+
+        shouldBeMissed = self.darkBBoxes[self.onlyEdges] + self.brightBBoxes[self.onlyEdges]
+        for defect in defects:
+            self.assertNotIn(defect.getBBox(), shouldBeMissed)
+
+    def test_postProcessDefectSets(self):
```

Comment:
This method does many things: please add a docstring saying what sorts of tests it does.

It wasn't clear to me how the different cases below actually test different aspects of the code, since all of the `asserts` are the same. What would make one of those asserts fail?

Reply:
Have documented this now. There is potential for logic errors when doing this combination step, so test it carefully for the different modes that are possible.